### ISSUE TITLE
Camera off when skipping config screen, despite set to "on" (first time, before permissions granted)

### ISSUE THREAD

**Issue by jgarplind** [Azure/communication-ui-library#4638]
**Describe the bug; what happened?**

The camera starts in an `off` state, despite explicitly requested to be on: `adapter.joinCall({ microphoneOn: true, cameraOn: true });`

**What are the steps to reproduce the issue?**

1. Join a call which runs the example code from (https://github.com/Azure/communication-ui-library/discussions/3379)
2. Get stuck in "Joining call..." until you accept the request for video/mic permissions
3. Now you are in the call, and the video is **off**, despite joining with `adapter.joinCall({ microphoneOn: true, cameraOn: true })`

This would typically just be a small nuissance, but we are trying to create a seamless low-interaction video experience where the users do not need to manage their own video, thus we have modified the call controls to use `cameraButton: false`, which means that the user cannot recover from this state without exiting the call.

On 2nd attempt and onwards, this is not a problem, because the permissions have already been granted.

**What behavior did you expect?**

Camera should be on when executing `adapter.joinCall({ microphoneOn: true, cameraOn: true })`, or perhaps a helpful error could be thrown to let application developers handle this scenario in user-land.

**If applicable, provide screenshots:**

N/A, please let me know if you think differently.

**In what environment did you see the issue?**

- `@azure/communication-react` npm package version (if applicable): 1.16.0
- `@azure/communication-calling` npm package version (if applicable): 1.24.1

- OS & Device: Multiple, doubt this is OS/Device-dependent, but at least Android (multiple devices) and Windows
- Browser: Google Chrome, Android WebView, probably more.
- Browser Version: >=124

**Is there any additional information?**

I tried various workarounds with timeouts, manually requesting permissions, and manually starting the camera.

What I finally got to work was this:

```diff
  const afterAdapterCreate = useCallback(
    async (adapter: CallAdapter): Promise<CallAdapter> => {
      adapter.joinCall({ microphoneOn: true, cameraOn: true });
+     adapter.onStateChange((state) => {
+       if (state.devices.cameras.length >= 1 && state.cameraStatus === "Off") {
+         adapter.startCamera();
+       }
+     });
      return new Promise((resolve) => resolve(adapter));
    },
    [],
  );
```

I'm not too keen on keeping `if (e.devices.cameras.length >= 1 && e.cameraStatus === "Off")` as a solution long-term. Ideally I think this caveat should be handled within the library, but it might unblock us for the time being.

**Comment by mgamis-msft** [comment]
Hi @jgarplind, thanks for filing this issue and detailing the repro steps. Also thank you for sharing your workaround. We will look into this and look for a proper fix in our library.

**Comment by mgamis-msft** [comment]
@jgarplind. I discussed this issue with my team and we determined that if developers want to join a call with devices on while skipping the config screen then they are responsible for ensuring the corresponding device permissions are granted.

This is a quick, though rather hacky, example of how to ensure camera devices are available before joining a call immediately with the camera on:
```
  const afterCallAdapterCreate = useCallback(
    async (adapter: CallAdapter): Promise<CallAdapter> => {
      subscribeAdapterEvents(adapter);
      adapter.askDevicePermission({ video: true, audio: true });
      let cameraPermissionState: PermissionState = (
        await navigator.permissions.query({ name: 'camera' as PermissionName })
      ).state;
      let microphonePermissionState: PermissionState = (
        await navigator.permissions.query({ name: 'microphone' as PermissionName })
      ).state;
      let cameras = await adapter.queryCameras();
      let microphones = await adapter.queryMicrophones();
      while (
        cameras.length === 0 ||
        microphones.length === 0 ||
        microphonePermissionState === 'prompt' ||
        cameraPermissionState === 'prompt'
      ) {
        cameraPermissionState = (await navigator.permissions.query({ name: 'camera' as PermissionName })).state;
        microphonePermissionState = (await navigator.permissions.query({ name: 'microphone' as PermissionName })).state;
        cameras = await adapter.queryCameras();
        microphones = await adapter.queryMicrophones();
        await new Promise((_) => setTimeout(_, 200));
      }
      adapter.joinCall({ microphoneOn: true, cameraOn: true });
      return adapter;
    },
    [subscribeAdapterEvents]
  );
  ```

The configuration screen handled requesting device permissions as well letting users know device permissions are needed through prompts before joining the call. So a more proper way would be to return a different page component before creating the adapter if device permissions are not yet granted. Something like:

```
adapter.askDevicePermission({ video: true, audio: true });
const cameraPermissionState: PermissionState = (
    await navigator.permissions.query({ name: 'camera' as PermissionName })
).state;
const microphonePermissionState: PermissionState = (
    await navigator.permissions.query({ name: 'microphone' as PermissionName })
).state;

if (cameraPermissionState !== 'granted' && microphonePermissionState !== 'granted') {
    <Some page component telling the camera and microphone permissions are required/>
}

const afterAdapterCreate = useCallback(
    async (adapter: CallAdapter): Promise<CallAdapter> => {
        adapter.joinCall({ microphoneOn: true, cameraOn: true });
        await adapter.queryCameras();
        await adapter.queryMicrophones();
        return new Promise((resolve) => resolve(adapter));
    },
    [],
);

return <Component page where adapter is created afterAdapterCreate={afterAdapterCreate} />;
```
The prompt components we use in our configuration page to notify user that camera and microphone permissions are needed are documented in this [storybook page](https://azure.github.io/communication-ui-library/?path=/docs/ui-components-site-permissions--site-permissions).

I hope this helps.

**Comment by mgamis-msft** [comment]
@jgarplind
- Yes those are two separate workarounds. The first ensures device permissions by ensure camera and mic permission are granted or rejected and that there are at least one device for each, though in a hacky way in a blocking while loop. The second is less blocking since it does not create the adapter and join the call before you ensure devices permissions are granted.
- `subscribeAdapterEvents` is just something we have in our sample Calling app to log some information to console. You can take a look [here](https://github.com/Azure/communication-ui-library/blob/main/samples/Calling/src/app/views/CallScreen.tsx#L47)
- A PR for better documentation would be very helpful, especially from an outside perspective. It would give a really great outlook of what we are missing and have overlooked. Your contribution would be much appreciated

**Comment by jgarplind** [comment]
Thanks for clarifying.

I'll try to get around to adding docs eventually. Would this qualify as a sample, or is there a better place to put information like this? The less friction and ambiguity, the more likely I'll actually get around to doing it.

Thanks again!

**Comment by mgamis-msft** [comment]
I see @jgarplind. You want to add a sample app that skips the config screen. We don't plan on changing our sample apps in this repo to skip the configuration screen. But it would be awesome, if you can share your github sample repo that ensure permissions and skips the config screen in a comment in the [skip config screen discussion](https://github.com/Azure/communication-ui-library/discussions/3379).
I guess we don't really have better place that talks about skipping configuration other than the discussion. But if you are keen on it you can add an mdx in our storybook [here](https://github.com/Azure/communication-ui-library/tree/main/packages/storybook/stories) or even just a snippet file showing skipping config screen with camera on and mic to help us get started.
Don't feel too much pressure into helping us with docs. Even just sharing a snippet in the discussion would be great.
I will close this issue. Feel free to create another issue if anything else comes up.
